File: file_entropy/entropy.go
package main

import (
	"fmt"
	"math"
	"os"
	"io/ioutil" )

func getByteEntropy(filename string) (float64, error) {
	bs, err := ioutil.ReadFile(filename)
	if err != nil {
		return 0.0, err
	}

	byte_counters := make(map[byte]int)
	for _, i := range bs {
		byte_counters[i] += 1
	}

	var entropy float64
	for _, count := range byte_counters {
		p_i := float64(count) / float64(len(bs))
		entropy -= p_i * math.Log2(p_i)
	}

	return entropy, nil
}

func main() {
	for _, element := range os.Args[1:] {
		entropy, err := getByteEntropy(element)
		if err == nil {
			fmt.Printf("%s\t%f\n", element, entropy)
		}
	}
}

File: file_entropy/entropy.rb
#!/usr/bin/ruby -w

# Function to calculate the byte-wise entropy of a file.
def getByteEntropy(filename)
begin
	contents = IO.binread(filename).split("")
	# Make a histrogram from the array 'contents'
	byte_counters = Hash[*contents.group_by{ |v| v }.flat_map{ |k, v| [k, v.size] }]

	entropy = 0.0
	total_size = contents.length().to_f
	byte_counters.each_value do |count|
		p_i = count / total_size
		entropy -= Math.log2(p_i) * p_i
	end
	return entropy.round(5)

rescue
	# Silently continue
end
end


# Main
ARGV.each do |argv|
	entropy = getByteEntropy(argv)
	if entropy
		puts argv + "\t" + entropy.to_s
	end
end
